## First project: Python Analysis on profitable App Profiles for App Store and Google Play Store

Hi everyone, this is my first project using Python to analyse the profitable app profiles for app store and google play store. In this scenario, I am working as a data analysts for a company that builds Android and iOS mobile apps.

In this company, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. We only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means that the number of users of our apps determines our revenue for any given app — the more users who see and engage with the ads, the better. Our goal for this project is to analyze data to help our developers understand what type of apps are likely to attract more users.

### Opening and exploring the Data sets

As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play.

* [The Google Play Store Dataset](https://www.kaggle.com/datasets/lava18/google-play-store-apps/data) contains data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018.

* [The Apple Store Dataset](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps) contains data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017.

Lets explore both the dataset by loading it into Jupyter Notebook.

We are going to load both the `'AppleStore.csv'` and `'googleplaystore.csv'` into Jupyter Notebook in the following code chunk:

1) Open the `'AppleStore.csv'` and `'googleplaystore.csv'` using the `open()` function
2) Import `reader()` function from the `csv` module
3) read the file using `reader()` function, and assigning the variable to `read_file`
4) open the `read_file` as a list-of-list using the `list()` function and named both as `apple_apps` and `google_apps` accordingly
5) we created separate list for the header and body of the dataset 

In [3]:
from csv import reader

#### This is App Store data set ####
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
apple_apps = list(read_file)
apple_header = apple_apps[0]
apple_apps = apple_apps[1:]

#### This is Andriod data set ####
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
google_apps = list(read_file)
google_header = google_apps[0]
google_apps = google_apps[1:]

In the Function below, we defined `explore_data()` function. The `explore_data()` function is used to inspect a specific portion of dataset and optionally view its structure.

In [5]:
def explore_data(data_set, start, end, no_rows_and_column = False):
    dataset_slice = data_set[start:end]
    for row in dataset_slice:
        print(row)
        print('\n') #adds an empty line after each row

    if no_rows_and_column:
        print('Number of rows:', len(data_set))
        print('Number of column:', len(data_set[0]))

In [6]:
print(apple_header)
print('\n')
explore_data(apple_apps,0,5,True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 7197
Number of column: 16


We have 7197 number of rows in iOS apps in the dataset, and the column names are not easily understandable, namely `'track_name'`, `'currency'`, `'rating_count_tot'`, '`rating_count_ver'`. Details of each columns can be found in dataset [documentation](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps)

In [8]:
print(google_header)
print('\n')
explore_data(google_apps,0,5,True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Eve

in the Andriod dataset, we have 10841 number of rows, and 13 columns. The columns that can be useful for our analysis could be `'App'`, `'Category'`, `'Rating'`, `'Reviews'`, `'Installs'`, `'Price'`,`'Content Rating'`, and `'Genres'`. The details of the column names can be found in the [documentation](https://www.kaggle.com/datasets/lava18/google-play-store-apps) of the dataset.

|Name in iOS| Name in Andriod| Description|
|-----------|----------------|------------|
|"id"| NULL| App ID|
|"track_name" | App| Application Name|
|"size_bytes" | Size| Size of app|
|"price" | Price| Price of app|
|"user_rating" | Rating | Overall user rating|
|"cont_rating" | Content Rating | Content Rating |
|"prime_genre" | Genres| Genres of app|
|NULL | Installs | Number of user downloads|
|NULL | Reviews | Number of user reviews|

I have labelled some column names in both the datasets for better understading of the datasets.

## Deleting Wrong Data

Now, we look for the number of columns in our dataset that does not match

In [14]:
for row in google_apps:
    header_length = len(google_header)
    if len(row) != header_length:
        ##Reason why not using len(google_apps), is because we are comparing header_length against single number of row, instaed of total number of rows in dataset
        print(row)
        print(google_apps.index(row))

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10472


In [15]:
print(google_apps[10472])
print('\n')
print(google_header)

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


The row 10472 corresponds to app *Life Made WI-Fi Touchscreen Photo Frame*, the value under `Rating` column is 19, which is clearly above the maximum rating for app at 5. This problem is caused by a missing value in the `Category` of the row. For this reason, we will delete the row.

In [17]:
print(len(google_apps))
del google_apps[10472]
print(len(google_apps))

10841
10840


## Removing duplicated entries

#### Part One

To find out what are the duplicated App in `google_apps` dataset. 

In [20]:
for app in google_apps:
    name = app[0]
    if name == "Instagram":
        print(google_header)
        print(app)
        print('\n')

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['App', 'Cate

There seems to be 4 entries of **Instagram** in the dataset.

Now we can create 2 seperate list to store the values of duplicated apps, and see how many duplicates we are dealing with

In [23]:
def duplicated_check(dataset, index):
    duplicated_apps = []
    unique_apps = []

    for row in dataset:
        name = row[index]
        if name in unique_apps:
        ## if the name has already listed in unique_apps, then all into duplicated apps
            duplicated_apps.append(name)
        else: 
            unique_apps.append(name)
        
    return duplicated_apps
# Check for duplicates in both the datasets
apple_duplicates = duplicated_check(apple_apps, 0)
google_duplicates = duplicated_check(google_apps, 0)

print('Number of Apple Duplicates:', len(apple_duplicates))
print('\n')
print('Example of Apple Duplicates:', (apple_duplicates)[:5])
print('\n')
print('Number of Google Duplicates:', len(google_duplicates))
print('\n')
print('Example of Google Duplicates:', (google_duplicates)[:5])

Number of Apple Duplicates: 0


Example of Apple Duplicates: []


Number of Google Duplicates: 1181


Example of Google Duplicates: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


we can see that there is **1,181** cases of duplicates in `google_apps` dataset

Based on the findings, we should remove the duplicate entries and keep only one entry per app. But we should look into which entry to remove. 
If we look into the **Instagram** duplicate entries above, we can see that there are differences between reviews. For the sake of accuracies and reliability, 
we will keep only the entries with highest number of reviews. 

#### Part Two

We will first start by building a dictionary. The purpose of the dictionary is to hold the unique app names with highest review ratings. We first start by building a loop. The `if` statement used to ensure that `reviews_max` always holds the highest number of reviews encountered for each app. It updates the value with the higher review count. The `elif` statement on the other hand, checks if the app name is not already a key in the `reviews_max` dictionary, and adds a new entry when it detects it. 







In [27]:
reviews_max = {}
for app in google_apps:
    name = app[0]
    n_reviews = float(app[3])

    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews

    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        

In a previous code cell, we found that there are 1,181 cases where an app occurs more than once, so the length of our dictionary (of unique apps) should be equal to the difference between the length of our data set and 1,180.

In [29]:
print('Expected length:', len(google_apps) - 1181)
print('actual length:', len(reviews_max))

Expected length: 9659
actual length: 9659


In [30]:
google_clean = []
already_added = []

for app in google_apps:
    name = app[0]
    n_reviews = float(app[3])
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        google_clean.append(app)
        already_added.append(name)

The loop will add the current row (app) into the list `google_clean` if the reviews of the app matches the description in `reviews_max`. The already added list accounts for the apps that have the same number of reviews with multiple entries.

In [32]:
explore_data(google_clean,0,5,True)
print('\n')
print('Actual Length:',len(reviews_max))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows: 9659
Number of column: 13


Actual Length: 9659


## Remove Non English apps
#### Part One

There are some names of the apps that are not written in english. Below, we can see a few example


In [34]:
print(apple_apps[813][1])
print(apple_apps[6731][1])
print('\n')
print(google_clean[4412][0])
print(google_clean[7940][0])
      

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜


中国語 AQリスニング
لعبة تقدر تربح DZ


We're not interested in keeping these apps, so we'll remove them. One way to do this is to remove each app with a name containing a symbol that isn't commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /).

In the function that we define below, we will use the `ord()` function to determine if the app name belongs to a set of common English character or not. According to [ASCII](https://en.wikipedia.org/wiki/ASCII), the characters that we commonly use in English text are all in the range 0 to 127. So, if an app name is greater than 127, then it probably means that the app has a non-English name. 

In [37]:
def is_english(string):
    for character in string:
        if ord(character) > 127:
            return False

    return True

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


In [38]:
print(ord('😜'))
print(ord('™'))

128540
8482


However, the function also detects character such as `™` or emoji like `😜` outside of the range of 127. In this case, we will need to refine the function to include these apps for better accuracies.

In [40]:
def is_english(string):
    non_english_count = 0
    
    for character in string:
        if ord(character) > 127:
            non_english_count += 1
            if non_english_count > 3:
                return False

    return True

print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

False
True
True


Below, we used the `is_english` function to determine the English apps for both dataset

In [42]:
google_english = []
apple_english = []

for app in google_clean:
    name = app[0]
    if is_english(name):
        ##it cant be "if is_english(name) in google_clean, as it tries to find Boolean value in the list
        google_english.append(app)

for app in apple_apps:
    name = app[1]
    if is_english(name):
        apple_english.append(app)

explore_data(google_english,0,5,True)
print('\n')
explore_data(apple_english,0,5,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows: 9614
Number of column: 13


['284882215', 'Facebook', '389879808', 'USD', '

## Isolating Free Apps

In the introduction, we have mentioned that the analysis only requires the free to download and install apps, and main source of revenue 

In [44]:
free_apple_apps = []
free_google_apps = []

for row in apple_english:
    price = row[4]
    if price == '0.0':
        free_apple_apps.append(row)

for row in google_english:
    price = row[7]
    if price == '0':
        free_google_apps.append(row)

explore_data(free_apple_apps,0,5,True)
print('\n')
explore_data(free_google_apps,0,5,True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 3222
Number of column: 16


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', '

## Analysis

### Most Common apps by Genre

#### Part One

As mentioned in the introduction, our goal is to determine the kinds of apps that are most likely to attract more users, as the number of people using the app directly correlates to the revenue. Therefore, we will look at successful app profiles that fit both the iOS and Andriod operating system, as it caters to more user base.

To minimize risks and overhead, our validation strategy for an app idea is comprised into three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Lets begin our analysis by inpecting both the dataset and identify columns that we could identify trends for most popular genres. For this, we'll build a frequency table for the `prime_genre` column of the App Store data set, and the `Genres` and `Category` columns of the Google Play data set.



#### Part Two
We'll build two functions we can use to analyze the frequency tables:

1. One function to generate frequency tables that show percentages
2. Another function that we can use to display the percentages in a descending order

In [47]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
            
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage
        ## table_percentages[key] stores the percentage calculation to its corresponding key back into the dictionary

    return table_percentages
            
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

#### Part Three

We will examine the frequency table for `prime_genre` column of the App Store data set.

In [49]:
print('PERCENTAGE OF "PRIME GENRE" IN APPLE APP STORE:'"\n")
display_table(free_apple_apps,(11))

PERCENTAGE OF "PRIME GENRE" IN APPLE APP STORE:

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


In Apple dataset, we could conclude that the most common free app in `Prime_genre` is *Games*. Alone, it already takes up 58% of the entire free English app market in iOS. The next most common genre is *Entertainment*, which takes up about 7.88%. At a glance, the top categories in `prime_genre` are leaning more towards entertainment, with (Games, photo and videos, social networking, sports), while apps with practical purposes like (education, shopping, utilities, productivity, lifestyle) are rare. However, eventhough entertaining apps seem to have a big share of the pie, it doesn't imply that number of users directly correlate to it. Any recommendations will need further analysis and investigation on the dataset. 

We will continue by investigating the `Category` and `Genre` columns of Google Play dataset below:

In [51]:
print('PERCENTAGE OF "CATEGORIES" IN GOOGLE PLAYSTORE:'"\n")
display_table(free_google_apps,(1))

PERCENTAGE OF "CATEGORIES" IN GOOGLE PLAYSTORE:

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661

In the first glance, the `Category` column of google play dataset has **Family** as the most common category, taking up 18% of the total column, with **Game** at 9.72% and **Tools** at 8.46% following closely behind. The landscape seems significantly different in Google Play dataset, with good number of apps designed for practical purposes(family, tools, business, lifestyle, productivity). However, Upon checking, the **Family** Category (Which is now the **Kids** section in google playstore) consist of only apps that are appropriate for kids, that are mainly games for kids.

![Kids Category](Apple_vs_Google_apps_Kids_Category.png)

But looking at its entirety, the practical apps still seems to have better representation on Google Playstore than on Apple App Store. Eventhough our frequency table above has revealed the most frequent app genres by categories, it does not discuss what genres have the most users. Therefore, further analysis is still required for a recommendation for app profile. 

In [53]:
print('PERCENTAGE OF "GENRES" IN GOOGLE PLAYSTORE:'"\n")
display_table(free_google_apps,(9))

PERCENTAGE OF "GENRES" IN GOOGLE PLAYSTORE:

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.93637

In `Genres` column of Google Play dataset, `Tools` takes the lead at 8.44%, while `Entertainment` is closely following behind. The percentage gap between these categories are relatively small. However, notice that the main categories are split into smaller groups of apps further down the list. For example, (*Educational;Creativity & Educational;Action & Adventure*), concluding that the `Genre` column is more detailed and specific. As we are not diving into the specifics in this study, we will move forward with just the `Category` column. 

## Most popular app by Genre in App Store

One way to find out the most popular app is to find out the number of users it has for each app category. However, the column is missing in the `AppleStore.csv dataset`. To workaround this, we could use the total number of user ratings as a proxy, which can be found in the `rating_count_tot` column. 

we could start by getting the average ratings for each app category as shown below:

In [56]:
genre_apple = freq_table(free_apple_apps, 11)
table_display = []        

##genre represents each key in genre_apple by default in python
for genre in genre_apple:
    total = 0
    len_genre = 0
    for app in free_apple_apps:
        genre_app = app[11]
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
            
    if len_genre > 0:
        avg_n_ratings = total / len_genre
        table_display.append((avg_n_ratings, genre))

# Sort the list by average ratings in descending order
table_sorted = sorted(table_display, reverse=True)

# Print each entry with the genre and average ratings
for entry in table_sorted:
    print(entry[1], ':', entry[0])

Navigation : 86090.33333333333
Reference : 74942.11111111111
Social Networking : 71548.34905660378
Music : 57326.530303030304
Weather : 52279.892857142855
Book : 39758.5
Food & Drink : 33333.92307692308
Finance : 31467.944444444445
Photo & Video : 28441.54375
Travel : 28243.8
Shopping : 26919.690476190477
Health & Fitness : 23298.015384615384
Sports : 23008.898550724636
Games : 22788.6696905016
News : 21248.023255813954
Productivity : 21028.410714285714
Utilities : 18684.456790123455
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Business : 7491.117647058823
Education : 7003.983050847458
Catalogs : 4004.0
Medical : 612.0


On average, the highest rating count by category is **Navigation, Reference and Social Networking**. Lets find out which apps contributed the ratings for these category. Below, we will further conduct analysis on these categories:

In [58]:
for app in free_apple_apps:
    if app[11] == 'Navigation':
        print(app[1], ':', app[5])

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


In [59]:
for app in free_apple_apps:
    if app[11] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


In [60]:
for app in free_apple_apps:
    if app[11] == 'Social Networking' and (float(app[5]) > 100000):
        print(app[1], ':', app[5])


Facebook : 2974676
Pinterest : 1061624
Skype for iPhone : 373519
Messenger : 351466
Tumblr : 334293
WhatsApp Messenger : 287589
Kik : 260965
ooVoo – Free Video Call, Text and Voice : 177501
TextNow - Unlimited Text + Calls : 164963
Viber Messenger – Text & Call : 164249
Followers - Social Analytics For Instagram : 112778


In Apple AppStore, the most user rated app is navigation with average number of rating at *86,090* and followed by Reference at *74,942*, and Social Networking at *71,548*. 

The top 2 in the first 3 of the list are more practical tools that are difficult to insert ads into. For instance, navigation apps emphasizes on user experience, so ads should not interfere with core functionality, such as real-time nagivation and map usage. Ads could also deemed as ineffective as navigation apps are not as interactive. We could also observe that the ratings in navigation category is highly skewed as most of the rating falls into the app: "Waze" and "Google Maps", while the apps below struggles to even get more than 10,000 ratings.

As for reference apps, users mainly use these apps to receive information or educational content, we need to ensure that ads do not distrupt the user's ability to access and consume content. Also, placing an ad in Reference type app would need for it to be relevant to the content or educational value of the app to avoid distruption of content consumption. I would also like to discuss on the average rating for reference apps in App Store. In this category, the same pattern from Navigation category could be observed, with "Bible" app being the outlier with over 900,000 ratings, which skew up the average rating.

In App Store, the average ratings on Social Networking app category seems heavily skewed with tech giants such as Facebook, Pinterest, Tumblr, Whatsapp and etc. Even though Social networking apps are highly interactive, with great users engagement, this category seems to be oversaturated and heavily infleunced by these app giants. 

However, the **Shopping** category seems to show some potential, as the nature of *shopping* category seems to fit into the criteria of being highly interactive and easily integrating ads into the apps. Lets take a look on the category. 

In [62]:
for app in free_apple_apps:
    if app[11] == 'Shopping' and (float(app[5]) > 100000):
        print(app[1], ':', app[5])

Groupon - Deals, Coupons & Discount Shopping App : 417779
eBay: Best App to Buy, Sell, Save! Online Shopping : 262241
Wish - Shopping Made Fun : 141960
shopkick - Shopping Rewards & Discounts : 130823
Amazon App: shop, scan, compare, and read reviews : 126312
Target : 108131
Zappos: shop shoes & clothes, fast free shipping : 103655


It looks like there is only a few popular apps, with popular apps' rating generally below 200,000. So, this market still shows some potential.

In essence, shopping app would provide the most value for in app ads. Ads can be easily integrated into shopping apps using formats like banner ads, interstitial ads, native ads, and video ads. These ads generate additional revenue, enhance user engagement, and can be tailored to match the app’s design and user preferences. In contrast, navigation and references apps require a more cautious approach to ensure ads do not distrupt the core functionalities of the app, while social networking apps seem to be oversaturated and dominated by tech giant apps at the moment. 

This idea of fitting ads into shopping app category may fit well as App Store is mainly dominated with for-fun apps. So a shopping app may seem to cater to the user base of App Store. Also, user base of shopping apps typically has a mindset oriented toward spending money when using the app. Users generally visit shopping apps with the intent to browse, compare, and purchase products, making them more receptive to spending. These apps are designed to facilitate and encourage purchases through features like personalized recommendations, discounts, and seamless checkout processes. wWen in app ads are done thoughtfully, integrating ads into shopping apps can enhance conversion rates and user retention by providing value-added content and personalized offers.

Now lets look at the popular app genre of Google Play Store. 

## Most popular app by Genre in Google Play Store

Now lets find out the most popular app in Google PlayStore. We have the number of users column under `Install` in the Google Playstore dataset. Let's check it out by using the `display_table()` function.

In [65]:
display_table(free_google_apps, 5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


However, the values given in the dataset is not absolute, and is also a string. For example, `1,000,000+` does not precisely list out the exact number of installs. But since we don't need very precise data for our purposes — we only want to find out which app genres attract the most users. This means that we'll need to convert each install number from a string to a float. This means we need to remove the commas and the plus characters, or the conversion will fail and cause an error.

In [67]:
google_category = freq_table(free_google_apps, 1)
google_display = []

for genre in google_category:
    total = 0
    len_category = 0
    for app in free_google_apps:
        category = app[1]
        if category == genre:
            n_installs = app[5]
            n_installs = n_installs.replace('+', '')
            n_installs = n_installs.replace(',', '')
            total += float(n_installs)
            len_category += 1
            
    if len_category > 0:
        avg_n_installs = total / len_category
        google_display.append((avg_n_installs, genre))

google_sorted = sorted(google_display, reverse = True)

for entry in google_sorted:
    print(entry[1], ':', entry[0])

COMMUNICATION : 38456119.167247385
VIDEO_PLAYERS : 24727872.452830188
SOCIAL : 23253652.127118643
PHOTOGRAPHY : 17840110.40229885
PRODUCTIVITY : 16787331.344927534
GAME : 15588015.603248259
TRAVEL_AND_LOCAL : 13984077.710144928
ENTERTAINMENT : 11640705.88235294
TOOLS : 10801391.298666667
NEWS_AND_MAGAZINES : 9549178.467741935
BOOKS_AND_REFERENCE : 8767811.894736841
SHOPPING : 7036877.311557789
PERSONALIZATION : 5201482.6122448975
WEATHER : 5074486.197183099
HEALTH_AND_FITNESS : 4188821.9853479853
MAPS_AND_NAVIGATION : 4056941.7741935486
FAMILY : 3695641.8198090694
SPORTS : 3638640.1428571427
ART_AND_DESIGN : 1986335.0877192982
FOOD_AND_DRINK : 1924897.7363636363
EDUCATION : 1833495.145631068
BUSINESS : 1712290.1474201474
LIFESTYLE : 1437816.2687861272
FINANCE : 1387692.475609756
HOUSE_AND_HOME : 1331540.5616438356
DATING : 854028.8303030303
COMICS : 817657.2727272727
AUTO_AND_VEHICLES : 647317.8170731707
LIBRARIES_AND_DEMO : 638503.734939759
PARENTING : 542603.6206896552
BEAUTY : 51315

The top 3 categories of most installed apps are **COMMUNICATION**, **VIDEO_PLAYERS**, and **SOCIAL**. However, these apps might be skewed as some of the biggest apps are in these categories. We will further analyse **COMMUNICATION** app category below:

In [69]:
for app in free_google_apps:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+' or
                                      app[5] == '500,000,000+' or
                                      app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

In [70]:
under_100m_communication = []
for app in free_google_apps:
    n_installs = app[5]
    n_installs = n_installs.replace('+', '')
    n_installs = n_installs.replace(',', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) <100000000):
        under_100m_communication.append(float(n_installs))

print('COMMUNICATION APPS AVERAGE UNDER 100M:',sum(under_100m_communication) / len (under_100m_communication))
print('\n')
first_entry = google_sorted[0]
print(first_entry[1], ':', first_entry[0])

COMMUNICATION APPS AVERAGE UNDER 100M: 3603485.3884615386


COMMUNICATION : 38456119.167247385


Even though the **COMMUNICATION** category has the highest average downloads, When results of installs over 100 millions are taken out, the average falls to *3.6 Million* from *38 Million* average downloads. This is due to apps like WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts gathering on average over 1 billion downloads, thus heavily skewing the result of average app downloads. 

Now, lets look at **VIDEO_PLAYERS** category.

In [72]:
for app in free_google_apps:
    if app[1] == 'VIDEO_PLAYERS' and (app[5] == '1,000,000,000+' or
                                      app[5] == '500,000,000+' or
                                      app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

YouTube : 1,000,000,000+
Motorola Gallery : 100,000,000+
VLC for Android : 100,000,000+
Google Play Movies & TV : 1,000,000,000+
MX Player : 500,000,000+
Dubsmash : 100,000,000+
VivaVideo - Video Editor & Photo Movie : 100,000,000+
VideoShow-Video Editor, Video Maker, Beauty Camera : 100,000,000+
Motorola FM Radio : 100,000,000+


In [73]:
under_100m_vid = []
for app in free_google_apps:
    n_installs = app[5]
    n_installs = n_installs.replace('+', '')
    n_installs = n_installs.replace(',', '')
    if (app[1] == 'VIDEO_PLAYERS') and (float(n_installs) <100000000):
        under_100m_vid.append(float(n_installs))

print('VIDEO_PLAYERS APPS AVERAGE UNDER 100M:',sum(under_100m_vid) / len (under_100m_vid))
print('\n')
first_entry = google_sorted[1]
print(first_entry[1], ':', first_entry[0])

VIDEO_PLAYERS APPS AVERAGE UNDER 100M: 5544878.133333334


VIDEO_PLAYERS : 24727872.452830188


As shown above, the deviation when apps above 100 million downloads (YouTube, Google Play Movies & TV, and MX Player) are taken out, the average falls to *5.5 Million* from *24 Million* which suggests that the result is skewed.

Now we look at **SOCIAL** category.

In [75]:
for app in free_google_apps:
    if app[1] == 'SOCIAL' and (app[5] == '1,000,000,000+' or
                                app[5] == '500,000,000+' or
                                app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Facebook : 1,000,000,000+
Facebook Lite : 500,000,000+
Tumblr : 100,000,000+
Pinterest : 100,000,000+
Google+ : 1,000,000,000+
Badoo - Free Chat & Dating App : 100,000,000+
Tango - Live Video Broadcast : 100,000,000+
Instagram : 1,000,000,000+
Snapchat : 500,000,000+
LinkedIn : 100,000,000+
Tik Tok - including musical.ly : 100,000,000+
BIGO LIVE - Live Stream : 100,000,000+
VK : 100,000,000+


In [76]:
under_100m_social = []
for app in free_google_apps:
    n_installs = app[5]
    n_installs = n_installs.replace('+', '')
    n_installs = n_installs.replace(',', '')
    if (app[1] == 'SOCIAL') and (float(n_installs) <100000000):
        under_100m_social.append(float(n_installs))

print('SOCIAL APPS AVERAGE UNDER 100M:',sum(under_100m_social) / len (under_100m_social))
print('\n')
first_entry = google_sorted[2]
print(first_entry[1], ':', first_entry[0])

SOCIAL APPS AVERAGE UNDER 100M: 3084582.5201793723


SOCIAL : 23253652.127118643


As suggested, the same pattern from **COMMUNICATION** and **VIDEO_PLAYERS** app can be seen for **SOCIAL** apps, which decreases to *3 Million* from *23 Million*. 

We see the same pattern for the top 3 categories. The market seems to be dominated by apps like Youtube, Google Play Movies & TV, or MX Player. The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

However, looking at the trend, the pattern for **SHOPPING** app category seems to be have a fair number of installs at 7,036,877. The shopping category is also a highly interactive app, with users who are looking to spend money. 

In [78]:
for app in free_google_apps:
    if app[1] == 'SHOPPING' and (app[5] == '1,000,000,000+' or 
                                      app[5] == '500,000,000+' or
                                      app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

eBay: Buy & Sell this Summer - Discover Deals Now! : 100,000,000+
Amazon Shopping : 100,000,000+
Flipkart Online Shopping App : 100,000,000+
AliExpress - Smarter Shopping, Better Living : 100,000,000+
Wish - Shopping Made Fun : 100,000,000+


However, it seems that only a few apps dominate in popularity, indicating that this market still has growth potential. Let's explore app ideas by focusing on those that fall in the mid-range of popularity, with downloads between 1,000,000 and 100,000,000.

In [80]:
for app in free_google_apps:
    if app[1] == 'SHOPPING' and (app[5] == '1,000,000+' or 
                                      app[5] == '5,000,000+' or
                                      app[5] == '10,000,000+' or
                                      app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Amazon for Tablets : 10,000,000+
OfferUp - Buy. Sell. Offer Up : 10,000,000+
Shopee - No. 1 Online Shopping : 10,000,000+
Shopee: No.1 Online Shopping : 10,000,000+
Kroger : 5,000,000+
Walmart : 10,000,000+
letgo: Buy & Sell Used Stuff, Cars & Real Estate : 50,000,000+
Lazada - Online Shopping & Deals : 50,000,000+
OLX - Buy and Sell : 50,000,000+
The wall : 1,000,000+
Flipp - Weekly Shopping : 10,000,000+
Shrimp skin shopping: spend less, buy better : 5,000,000+
Lotte Home Shopping LOTTE Homeshopping : 5,000,000+
Horn, free country requirements : 1,000,000+
Jiji.ng : 1,000,000+
GS SHOP : 10,000,000+
The birth : 50,000,000+
Home & Shopping - Only in apps. 10% off + 10% off : 10,000,000+
EHS Dongsen Shopping : 1,000,000+
bigbasket - online grocery : 5,000,000+
Bukalapak - Buy and Sell Online : 10,000,000+
Life market : 1,000,000+
Jabong Online Shopping App : 10,000,000+
Family Dollar : 1,000,000+
Jumia online shopping : 10,000,000+
Mercado Libre: Find your favorite brands : 50,000,000+


It looks like the category is pretty saturated in mid-range popularity with many e-commerce app achieving over 1 million downloads. Even though the competition might be significant, but it could also mean that there is demand for apps in the shopping category. It probably is a good idea to build an app in **shopping** category based solely on our analysis of the dataset. However, we will need some additional features such as integrating gamification elements, AI-powered personalization algorithm, in the app to stand out in this market. 

## Conclusion

We have analysed both the App Store and Google Play Store app dataset with recommending an app profile that can be profitable for both markets as goal in mind. 

We have concluded that getting into *Shopping* category may be profitable for both the App Store and Google Play Store markets. Even though this category may be over saturated with rise of new e-commerce apps, but with additional features such as integrating gamification elements, AI-powered personalization algorithm the app may be able to achieve success. 